activemodel/lib/active_model/type/helpers/numeric.rb

```diff
@@ -25,10 +25,18 @@ def cast(value)
         end
 
         def changed?(old_value, _new_value, new_value_before_type_cast) # :nodoc:
+          return false if unchanged_nan?(old_value, new_value_before_type_cast)
```

Comment:
Since `NaN` values are rare, I expect this branch to be often not taken, so I think we'd be better to check it last:

```ruby
(super || number_to_non_number) && !equal_nan?
```